// react-fuzzy/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _fuse = require('fuse.js');

var _fuse2 = _interopRequireDefault(_fuse);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var styles = {
  searchBoxStyle: {
    border: '1px solid #eee',
    borderRadius: 2,
    padding: '8px 10px',
    lineHeight: '24px',
    width: '100%',
    outline: 'none',
    fontSize: 16,
    color: '#666',
    boxSizing: 'border-box',
    fontFamily: 'inherit'
  },
  searchBoxWrapper: {
    padding: '4px',
    boxShadow: '0 4px 15px 4px rgba(0,0,0,0.2)',
    borderRadius: 2,
    backgroundColor: '#fff'
  },
  resultsStyle: {
    backgroundColor: '#fff',
    position: 'relative',
    padding: '12px',
    borderTop: '1px solid #eee',
    color: '#666',
    fontSize: 14
  },
  selectedResultStyle: {
    backgroundColor: '#f9f9f9',
    position: 'relative',
    padding: '12px',
    borderTop: '1px solid #eee',
    color: '#666',
    fontSize: 14
  },
  resultsWrapperStyle: {
    width: '100%',
    boxShadow: '0px 12px 30px 2px rgba(0, 0, 0, 0.1)',
    border: '1px solid #eee',
    borderTop: 0,
    boxSizing: 'border-box',
    maxHeight: 400,
    overflow: 'auto',
    position: 'relative'
  }
};

var defaultResultsTemplate = function defaultResultsTemplate(props, state, styl) {
  return state.results.map(function (val, i) {
    var style = state.selectedIndex === i ? styl.selectedResultStyle : styl.resultsStyle;
    return _react2.default.createElement(
      'div',
      { key: i, style: style },
      val.title
    );
  });
};

var FuzzySearch = function (_Component) {
  (0, _inherits3.default)(FuzzySearch, _Component);

  function FuzzySearch(props) {
    (0, _classCallCheck3.default)(this, FuzzySearch);

    var _this = (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(FuzzySearch).call(this, props));

    _this.state = {
      results: [],
      selectedIndex: 0
    };
    _this.handleChange = _this.handleChange.bind(_this);
    _this.handleKeyDown = _this.handleKeyDown.bind(_this);
    _this.fuse = new _fuse2.default(props.list, _this.getOptions());
    return _this;
  }

  (0, _createClass3.default)(FuzzySearch, [{
    key: 'getOptions',
    value: function getOptions() {
      var _props = this.props;
      var caseSensitive = _props.caseSensitive;
      var id = _props.id;
      var include = _props.include;
      var keys = _props.keys;
      var shouldSort = _props.shouldSort;
      var sortFn = _props.sortFn;
      var tokenize = _props.tokenize;
      var verbose = _props.verbose;
      var maxPatternLength = _props.maxPatternLength;
      var distance = _props.distance;
      var threshold = _props.threshold;
      var location = _props.location;


      return {
        caseSensitive: caseSensitive,
        id: id,
        include: include,
        keys: keys,
        shouldSort: shouldSort,
        sortFn: sortFn,
        tokenize: tokenize,
        verbose: verbose,
        maxPatternLength: maxPatternLength,
        distance: distance,
        threshold: threshold,
        location: location
      };
    }
  }, {
    key: 'getResultsTemplate',
    value: function getResultsTemplate() {
      var _this2 = this;

      return this.state.results.map(function (val, i) {
        var style = _this2.state.selectedIndex === i ? styles.selectedResultStyle : styles.resultsStyle;
        return _react2.default.createElement(
          'div',
          { key: i, style: style },
          val.title
        );
      });
    }
  }, {
    key: 'handleChange',
    value: function handleChange(e) {
      this.setState({
        results: this.fuse.search(e.target.value).slice(0, this.props.maxResults - 1)
      });
    }
  }, {
    key: 'handleKeyDown',
    value: function handleKeyDown(e) {
      var _state = this.state;
      var results = _state.results;
      var selectedIndex = _state.selectedIndex;

      if (e.keyCode === 40 && selectedIndex < results.length - 1) {
        this.setState({
          selectedIndex: selectedIndex + 1
        });
      } else if (e.keyCode === 38 && selectedIndex > 0) {
        this.setState({
          selectedIndex: selectedIndex - 1
        });
      } else if (e.keyCode === 13) {
        if (results[selectedIndex]) {
          this.props.onSelect(results[this.state.selectedIndex]);
        }
        this.setState({
          results: [],
          selectedIndex: 0
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props;
      var className = _props2.className;
      var width = _props2.width;
      var resultsTemplate = _props2.resultsTemplate;
      var placeholder = _props2.placeholder;
      var autoFocus = _props2.autoFocus;


      var mainClass = (0, _classnames2.default)('react-fuzzy-search', className);

      return _react2.default.createElement(
        'div',
        {
          className: mainClass,
          style: { width: width },
          onKeyDown: this.handleKeyDown
        },
        _react2.default.createElement(
          'div',
          { style: styles.searchBoxWrapper },
          _react2.default.createElement('input', {
            type: 'text',
            style: styles.searchBoxStyle,
            onChange: this.handleChange,
            ref: 'searchBox',
            placeholder: placeholder,
            autoFocus: autoFocus
          })
        ),
        this.state.results && this.state.results.length > 0 && _react2.default.createElement(
          'div',
          { style: styles.resultsWrapperStyle },
          resultsTemplate(this.props, this.state, styles)
        )
      );
    }
  }]);
  return FuzzySearch;
}(_react.Component);

exports.default = FuzzySearch;


FuzzySearch.propTypes = {
  caseSensitive: _react.PropTypes.bool,
  className: _react.PropTypes.string,
  distance: _react.PropTypes.number,
  id: _react.PropTypes.string,
  include: _react.PropTypes.array,
  maxPatternLength: _react.PropTypes.number,
  onSelect: _react.PropTypes.func.isRequired,
  width: _react.PropTypes.number,
  keys: _react.PropTypes.oneOfType([_react.PropTypes.array, _react.PropTypes.string]),
  list: _react.PropTypes.array.isRequired,
  location: _react.PropTypes.number,
  placeholder: _react.PropTypes.string,
  resultsTemplate: _react.PropTypes.func,
  shouldSort: _react.PropTypes.bool,
  sortFn: _react.PropTypes.func,
  threshold: _react.PropTypes.number,
  tokenize: _react.PropTypes.bool,
  verbose: _react.PropTypes.bool,
  autoFocus: _react.PropTypes.bool,
  maxResults: _react.PropTypes.number
};

FuzzySearch.defaultProps = {
  caseSensitive: false,
  distance: 100,
  include: [],
  location: 0,
  width: 430,
  placeholder: 'Search',
  resultsTemplate: defaultResultsTemplate,
  shouldSort: true,
  sortFn: function sortFn(a, b) {
    return a.score - b.score;
  },

  threshold: 0.6,
  tokenize: false,
  verbose: false,
  autoFocus: false,
  maxResults: 10
};

exports.default = FuzzySearch;
