// react-komposer/dist/compose.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.default = compose;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

var _shallowequal = require('shallowequal');

var _shallowequal2 = _interopRequireDefault(_shallowequal);

var _utils = require('./utils');

var _common_components = require('./common_components');

var _ = require('./');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function compose(fn, L1, E1) {
  var _ref = arguments.length <= 3 || arguments[3] === undefined ? {} : arguments[3];

  var contextTypes = _ref.contextTypes;
  var _ref$pure = _ref.pure;
  var pure = _ref$pure === undefined ? true : _ref$pure;
  var _ref$withRef = _ref.withRef;
  var withRef = _ref$withRef === undefined ? false : _ref$withRef;

  return function (ChildComponent, L2, E2) {
    (0, _invariant2.default)(Boolean(ChildComponent), 'Should provide a child component to build the higher order container.');

    if ((0, _utils.isReactNative)()) {
      (0, _invariant2.default)(L1 || L2, 'Should provide a loading component in ReactNative.');

      (0, _invariant2.default)(E1 || E2, 'Should provide a error handling component in ReactNative.');
    }

    var LoadingComponent = L1 || L2 || (0, _._getDefaultLoadingComponent)();
    var ErrorComponent = E1 || E2 || (0, _._getDefaultErrorComponent)();

    // If this is disabled, we simply need to return the DummyComponent
    if ((0, _.getDisableMode)()) {
      return (0, _utils.inheritStatics)(_common_components.DummyComponent, ChildComponent);
    }

    var Container = function (_React$Component) {
      (0, _inherits3.default)(Container, _React$Component);

      function Container(props, context) {
        (0, _classCallCheck3.default)(this, Container);

        var _this = (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Container).call(this, props, context));

        _this.getWrappedInstance = _this.getWrappedInstance.bind(_this);

        _this.state = {};

        // XXX: In the server side environment, we need to
        // stop the subscription right away. Otherwise, it's a starting
        // point to huge subscription leak.
        _this._subscribe(props, context);
        return _this;
      }

      (0, _createClass3.default)(Container, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
          this._mounted = true;
        }
      }, {
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(props, context) {
          this._subscribe(props, context);
        }
      }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
          this._mounted = false;
          this._unsubscribe();
        }
      }, {
        key: 'shouldComponentUpdate',
        value: function shouldComponentUpdate(nextProps, nextState) {
          if (!pure) {
            return true;
          }

          return !(0, _shallowequal2.default)(this.props, nextProps) || this.state.error !== nextState.error || !(0, _shallowequal2.default)(this.state.payload, nextState.payload);
        }
      }, {
        key: 'getWrappedInstance',
        value: function getWrappedInstance() {
          (0, _invariant2.default)(withRef, 'To access the wrapped instance, you need to specify ' + '{ withRef: true } as the fourth argument of the compose() call.');
          return this.refs.wrappedInstance;
        }
      }, {
        key: 'render',
        value: function render() {
          var error = this._getError();
          var loading = this._isLoading();

          if (error) {
            return _react2.default.createElement(ErrorComponent, { error: error });
          }

          if (loading) {
            return _react2.default.createElement(LoadingComponent, this._getProps());
          }

          return _react2.default.createElement(ChildComponent, this._getProps());
        }
      }, {
        key: '_subscribe',
        value: function _subscribe(props, context) {
          var _this2 = this;

          this._unsubscribe();

          var onData = function onData(error, payload) {
            if (error) {
              (0, _invariant2.default)(error.message && error.stack, 'Passed error should be an instance of an Error.');
            }

            var state = { error: error, payload: payload };

            if (_this2._mounted) {
              _this2.setState(state);
            } else {
              _this2.state = state;
            }
          };

          this._stop = fn(props, onData, context);
        }
      }, {
        key: '_unsubscribe',
        value: function _unsubscribe() {
          if (this._stop) {
            this._stop();
          }
        }
      }, {
        key: '_getProps',
        value: function _getProps() {
          var _state$payload = this.state.payload;
          var payload = _state$payload === undefined ? {} : _state$payload;


          var props = (0, _extends3.default)({}, this.props, payload);

          if (withRef) {
            props.ref = 'wrappedInstance';
          }

          return props;
        }
      }, {
        key: '_getError',
        value: function _getError() {
          var error = this.state.error;

          return error;
        }
      }, {
        key: '_isLoading',
        value: function _isLoading() {
          var payload = this.state.payload;

          return !Boolean(payload);
        }
      }]);
      return Container;
    }(_react2.default.Component);

    Container.contextTypes = contextTypes;
    return (0, _utils.inheritStatics)(Container, ChildComponent);
  };
}
